### PR TITLE
Add alb.targetGroupPrefix

### LINKED ISSUE #7292: ALB event Target Group names should have a prefix and not be completely random
# Feature Proposal

## Description

Context:
I see the issue: https://github.com/serverless/serverless/issues/6320 and the original implementation is to use the `<lambda-function-name>-<stage>` and this resulted in non-unique names for target groups.

I also see the fix: https://github.com/serverless/serverless/pull/6322 to simply use a fixed length hash to generate the target group's name.

Problem:
The problem with this approach is that we no longer have (semi) predictable names. From a resource management point of view, an AWS account admin would require additional AWS API calls to retrieve target group tags before he/she can manage them properly. In a large AWS account, these additional API calls may amount to significant costs and even account-wide API call throttling.

The second problem with this approach is that we cannot give granular permissions to our user/codebuild continuous deployment to only create or delete target groups for restricted names.

eg. a target group arn created by serverless looks like: `arn:aws:elasticloadbalancing:us-east-1:<redacted>:targetgroup/9ff09e8a342a4186247803c28a7b67a3/f7eb2224b059657e`

This means that I MUST give permissions of `*`.

```
  statement {
    actions = [
      "elasticloadbalancing:CreateTargetGroup",
      "elasticloadbalancing:DeleteTargetGroup",
    ]

    resources = [
      "arn:aws:elasticloadbalancing:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:targetgroup/*",
    ]
  }
```

This will not sit well with security audits. They would prefer a more restrictive policy like

```
"arn:aws:elasticloadbalancing:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:targetgroup/${my_resource_prefix}*"
```

Proposal:
Introduce an `alb.targetPrefix` to be prepended to your fixed length hash. Total length should still be truncated to 32 characters and `alb.targetPrefix` should have a max length of... maybe 10 or 15 characters.

Similar or dependent issues:

- https://github.com/serverless/serverless/pull/6322
- https://github.com/serverless/serverless/issues/6320

### PR DESCRIPTION
**Pull request by isen-ng**
## What did you implement

Added `provider.alb.targetGroupPrefix` parameter. When a value is provided, the value will be prepended to targetGroup created by the serverless framework. This allows users to create semi predictable target group names.

Closes #7292

## How can we verify it

```
service: my-service

provider:
  name: aws
  runtime: dotnetcore3.0
  stage: dev
  deploymentBucket: my-bucket
  alb:
    targetGroupPrefix: my-prefix

package:
  individually: true

functions:
  default:
    handler: not-required
    runtime: provided
    package:
      artifact: bin/release/netcoreapp3.0/my-service.zip
    events:
      - alb:
          listenerArn: my-alb-listener-arn
          priority: 100
          conditions:
            path: /*
```

## Todos

<details>
<summary>Useful Scripts</summary>

- [x] `npm run test:ci` --> Run all validation checks on proposed changes
- [x] `npm run lint:updated` --> Lint all the updated files
- [x] `npm run lint:fix` --> Automatically fix lint problems (if possible)
- [x] `npm run prettier-check:updated` --> Check if updated files adhere to Prettier config
- [x] `npm run prettify:updated` --> Prettify all the updated files

</details>

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** Yes
**_Is it a breaking change?:_** NO

### CODE REVIEW

**Comment by isen-ng** [comment]
How does this look?
i'll begin writing docs if this proposed addition fits your requirements

File: lib/plugins/aws/provider/awsProvider.js [hunk 1]
```diff
@@ -498,6 +498,25 @@ class AwsProvider {
...
   }
 
+  getAlbTargetGroupPrefix() {
+    const provider = this.serverless.service.provider;
+    if (
+      provider.alb === null ||
+      provider.alb === undefined ||
+      provider.alb.targetGroupPrefix === null ||
+      provider.alb.targetGroupPrefix === undefined
+    ) {
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js]: Let's simplify condition to `if (!provider.alb || !provider.alb.targetGroupPrefix)`

File: lib/plugins/aws/provider/awsProvider.js [hunk 2]
```diff
@@ -498,6 +498,25 @@ class AwsProvider {
...
+      return '';
+    }
+
+    if (provider.alb.targetGroupPrefix.length > 16) {
+      const errorMessage = `Length of alb.targetGroupPrefix should be at most 16 but is ${provider.alb.targetGroupPrefix.length}`;
+      throw new this.serverless.classes.Error(errorMessage);
+    }
+
+    return `${provider.alb.targetGroupPrefix}`;
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js]: No need for template wrap

File: lib/plugins/aws/lib/naming.js [hunk 3]
```diff
@@ -400,10 +400,13 @@ module.exports = {
...
   },
   getAlbTargetGroupName(functionName, albId, multiValueHeaders) {
-    return crypto
+    const hash = crypto
       .createHash('md5')
       .update(this.getAlbTargetGroupNameTagValue(functionName, albId, multiValueHeaders))
       .digest('hex');
+
+    const albTargetGroupName = this.provider.getAlbTargetGroupPrefix() + hash;
+    return albTargetGroupName.length > 32 ? albTargetGroupName.substr(0, 32) : albTargetGroupName;
```
Comment (medikoo) [inline lib/plugins/aws/lib/naming.js]: `substr` is deprecated, let's use `slice`

Comment (isen-ng) [reply lib/plugins/aws/lib/naming.js]: Switched to slice

File: lib/plugins/aws/provider/awsProvider.js (hunk 2 above)
Comment (isen-ng) [reply lib/plugins/aws/provider/awsProvider.js]: Removed template wrapping

File: lib/plugins/aws/provider/awsProvider.js (hunk 1 above)
Comment (isen-ng) [reply lib/plugins/aws/provider/awsProvider.js]: Suggestion adopted

**Comment by isen-ng** [comment]
@medikoo 
Thank you for your comments. I have implemented your suggestion.

In addition, I have also tried to update the documentation wherever I think I should. I would like your review on that as well. If there is anywhere else that needs updating, please point me in the right direction.

File: docs/providers/aws/events/alb.md [hunk 4]
```diff
@@ -103,3 +103,27 @@ module.exports.hello = async (event, context, callback) => {
 ```
 
 The handler response object must use `multiValueHeaders` to set HTTP response headers, `headers` would be ignored.
+
+## Prepending a prefix to generated target group names
+
+By default, target group names are strings generated by hashing a combined string of the function name, alb's id and whether multi-value headers are enabled. This produced a fixed length, unique id for target groups.
```
Comment (medikoo) [inline docs/providers/aws/events/alb.md]: _this produces_

File: docs/providers/aws/events/alb.md [hunk 5]
```diff
@@ -103,3 +103,27 @@ module.exports.hello = async (event, context, callback) => {
...
+
+Set the `providers.alb.targetGroupPrefix` attribute to a string to prepend this string to all target group names so that they become semi-predictable. The overall length is still truncated to 32 because this is a hard limit for target group names on AWS.
```
Comment (medikoo) [inline docs/providers/aws/events/alb.md]: _If you need target group names to have common predictable prefix, you may preset the prefix with `providers.alb.targetGroupPrefix` setting.  Note maximum length of the prefix is 16 chars_

File: docs/providers/aws/events/alb.md [hunk 6]
```diff
@@ -103,3 +103,27 @@ module.exports.hello = async (event, context, callback) => {
...
+
+Because of AWS's hard limit, `providers.alb.targetGroupPrefix` will also be limited to 16 characters in length to give ample length to the hash to ensure a high degree of uniqueness.
```
Comment (medikoo) [inline docs/providers/aws/events/alb.md]: Assuming we agree, on what I proposed above, let's remove this sentence

File: lib/plugins/aws/lib/naming.js [hunk 7]
```diff
@@ -400,10 +400,13 @@ module.exports = {
...
   },
   getAlbTargetGroupName(functionName, albId, multiValueHeaders) {
-    return crypto
+    const hash = crypto
       .createHash('md5')
       .update(this.getAlbTargetGroupNameTagValue(functionName, albId, multiValueHeaders))
       .digest('hex');
+
+    const albTargetGroupName = this.provider.getAlbTargetGroupPrefix() + hash;
+    return albTargetGroupName.length > 32 ? albTargetGroupName.slice(0, 32) : albTargetGroupName;
```
Comment (medikoo) [inline lib/plugins/aws/lib/naming.js]: It can be just `return albTargetGroupName.slice(0, 32)` (there's no harm on running `slice` on shorter strings)

Comment (isen-ng) [reply lib/plugins/aws/lib/naming.js]: Adopted your suggestion

File: docs/providers/aws/events/alb.md (hunk 6 above)
Comment (isen-ng) [reply docs/providers/aws/events/alb.md]: Removed

File: docs/providers/aws/events/alb.md (hunk 5 above)
Comment (isen-ng) [reply docs/providers/aws/events/alb.md]: Updated with minor changes

File: docs/providers/aws/events/alb.md (hunk 4 above)
Comment (isen-ng) [reply docs/providers/aws/events/alb.md]: Adopted entirely

**Comment by isen-ng** [comment]
Btw, is it possible to also give credit to @jinhong- for this PR?
He's a colleague of mine who wrote the initial parts of this PR. I merely filled out the tests and the docs.

**Comment by medikoo** [comment]
> Btw, is it possible to also give credit to @jinhong- for this PR?

Definitely, we plan to shortly issue some blog post listing latest contributions and we'll ensure to have @jinhong-  also there

**Pull request merged**